packages/mui-base/src/unstable_useNumberInput/useNumberInput.types.ts

```diff
@@ -1,10 +1,43 @@
 import * as React from 'react';
 import { FormControlState } from '../FormControl';
+import { NumberInputAction } from './numberInputAction.types';
+import { ActionWithContext } from '../utils/useControllableReducer.types';
 
-export type UseNumberInputChangeHandler = (
-  e: React.KeyboardEvent<HTMLInputElement>,
-  value: number | null,
-) => void;
+export type StepDirection = 'up' | 'down';
+
+/**
+ * The internal state of the NumberInput.
+ * Modify via the reducer only.
+ */
+export interface NumberInputState {
+  /**
+   * The clamped `value` of the `input` element.
+   */
+  value?: number | '';
+  /**
+   * The dirty `value` of the `input` element when it is in focus.
+   */
+  inputValue?: string;
```

Comment:
I'm leaning towards calling this `inputValue` again, it's easier to associate with `onInputChange` (even though this is just internal)